tests/e2e/stateapp/stateapp_test.go

```diff
@@ -0,0 +1,297 @@
+		nil,
+	}
+
+	var testCase1Key = guuid.New().String()
+	var testCase1Value = "The best song ever is 'Highwayman' by 'The Highwaymen'."
+
+	var testCaseManyKeys = generateRandomKeys(testManyEntriesCount)
+	var testCaseManyKeyValues = generateRandomValues(testCaseManyKeys)
+
+	return []testCase{
+		{
+			// No comma since this will become the name of the test without spaces.
+			"Test get save delete with empty request response",
+			[]testStep{
+				{
+					"get",
+					emptyRequest,
+					emptyResponse,
+				},
+				{
+					"save",
+					emptyRequest,
+					emptyResponse,
+				},
+				{
+					"delete",
+					emptyRequest,
+					emptyResponse,
+				},
+			},
+		},
+		{
+			// No comma since this will become the name of the test without spaces.
+			"Test save get and delete a single item",
+			[]testStep{
+				{
+					"get",
+					newRequest(simpleKeyValue{testCase1Key, nil}),
+					newResponse(simpleKeyValue{testCase1Key, nil}),
+				},
+				{
+					"save",
+					newRequest(simpleKeyValue{testCase1Key, &testCase1Value}),
+					emptyResponse,
+				},
+				{
+					"get",
+					newRequest(simpleKeyValue{testCase1Key, nil}),
+					newResponse(simpleKeyValue{testCase1Key, &testCase1Value}),
+				},
+				{
+					"delete",
+					newRequest(simpleKeyValue{testCase1Key, nil}),
+					emptyResponse,
+				},
+				{
+					"get",
+					newRequest(simpleKeyValue{testCase1Key, nil}),
+					newResponse(simpleKeyValue{testCase1Key, nil}),
+				},
+			},
+		},
+		{
+			// No comma since this will become the name of the test without spaces.
+			"Test save get and delete on multiple items",
+			[]testStep{
+				{
+					"get",
+					newRequest(testCaseManyKeys...),
+					newResponse(testCaseManyKeys...),
+				},
+				{
+					"save",
+					newRequest(testCaseManyKeyValues...),
+					emptyResponse,
+				},
+				{
+					"get",
+					newRequest(testCaseManyKeys...),
+					newResponse(testCaseManyKeyValues...),
+				},
+				{
+					"delete",
+					newRequest(testCaseManyKeys...),
+					emptyResponse,
+				},
+				{
+					"get",
+					newRequest(testCaseManyKeys...),
+					newResponse(testCaseManyKeys...),
+				},
+			},
+		},
+	}
+}
+
+var tr *runner.TestRunner
+
+func TestMain(m *testing.M) {
+	// These apps will be deployed for hellodapr test before starting actual test
+	// and will be cleaned up after all tests are finished automatically
+	testApps := []kube.AppDescription{
+		{
+			AppName:        appName,
+			DaprEnabled:    true,
+			ImageName:      "e2e-stateapp",
+			Replicas:       1,
+			IngressEnabled: true,
+		},
+	}
+
+	tr = runner.NewTestRunner(appName, testApps, nil)
+	os.Exit(tr.Start(m))
+}
+
+func TestServiceInvocation(t *testing.T) {
+	externalURL := tr.Platform.AcquireAppExternalURL(appName)
+	require.NotEmpty(t, externalURL, "external URL must not be empty!")
+	var testCases = generateTestCases()
+
+	// First we make some dummy calls to wait until the app is running.
+	// If this call fails because due to times out, it is OK.
```

Comment:
> it is OK

Can you clarify this?  It looks like if the get calls in HTTPGetNTimes() ever fail it'll return the first error, then the require.NoError() will fail.

Reply:
It is supposed to return the last error, not the first. Did you see a bug in HTTPGetNTimes()?

Reply:
Ok, that answers one question, I read it wrong [=

I was also asking about the comment "If this call fails because due to times out, it is OK.", though.  If it times out 3 times and returns an error indicting timeout then the require.NoError() will fail.  Isn't this behavior different than the comment suggests?

Reply:
The comment was for an older version of the code. Let me update that before approval of this PR.

Reply:
Updated comment.